Generate resource type filters from data

drewnoakes opened the pull request:
Note that currently that data is hard coded, but the new upcoming API work will allow the server to tell the client what types to expect.

drewnoakes committed:
Generate resource type filters from data

Note that currently that data is hard coded, but the new upcoming API work will allow the server to tell the client what types to expect.

drewnoakes commented on src/Aspire.Dashboard/Extensions/StringComparers.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace Aspire.Dashboard.Extensions;
+
+internal static class StringComparers
+{
+    public static StringComparer ResourceType => StringComparer.Ordinal;
+}
```
This is a convention I've seen used in a few places, giving semantic names to different domains of strings, ensuring they're compared in a uniform way.

drewnoakes commented on src/Aspire.Dashboard/Components/Pages/Resources.razor:
```diff
@@ -18,12 +18,17 @@
-                    <FluentCheckbox @bind-Value="_areProjectsVisible" Label="Projects" @bind-Value:after="HandleTypeFilterTypeChanged" />
-                    <FluentCheckbox @bind-Value="_areContainersVisible" Label="Containers" @bind-Value:after="HandleTypeFilterTypeChanged" />
-                    <FluentCheckbox @bind-Value="_areExecutablesVisible" Label="Executables" @bind-Value:after="HandleTypeFilterTypeChanged" />
+                    <FluentCheckbox
```
I had the impression it wasn't actually showing intederminate state in the previous code because of `ShowIndeterminate="false"`, but maybe that's just clearing indeterminate state if present.

I will reinstate the indeterminate state.

tlmii replied:
Yeah, the naming is a little confusing. `ShowIndeterminate` actually controls whether the user can toggle between on/off and indeterminate with mouse clicks. We don't want that, so it's false. `ThreeState` parameter controls whether the component will even display the three states in any situation (including programmatic, which we use), so that one was true.

You can see some docs and examples here if it helps: https://www.fluentui-blazor.net/Checkbox#fluentcheckbox-class_properties

tlmii replied:
To clarify the existing behavior:

The "All" checkbox would automatically toggle to checked (if all other checkboxes were checked), unchecked (if no other checkboxes were checked) or indeterminate (any other combination of checks). 
The user clicking it wouldn't ever toggle it to indeterminate, just between checked/unchecked - and those actions would impact all the other checkboxes as appropriate.

drewnoakes replied:
I checked the behaviour on `main`. Here are some recordings:

![tri-state-before](https://github.com/dotnet/aspire/assets/350947/5cb738de-ce4f-4373-b9fd-6132cbdf2895)

When starting from no selections and adding, it appears indeterminate:

![tri-state-before-2](https://github.com/dotnet/aspire/assets/350947/0d0a3907-d7ee-4c5d-8eab-c659c5294e88)

drewnoakes replied:
I played around with this for a while and am can't tell why it wouldn't work. Sticking a breakpoint in the property getter the `CheckState` value is bound to shows it's getting a `null` but displaying as a `false`. Anyone got any thoughts or ideas here?

drewnoakes committed:
Reinstate indeterminate state

Note that this is not yet working correcetly due to an apparent bug in the control being used. That's being investigated separately.

drewnoakes committed:
Merge remote-tracking branch 'upstream/main' into dynamic-resource-types

drewnoakes committed:
Change ShowIndeterminate value

tlmii commented:
> GIF of latest revision's behaviour:

For whenever someone is looking through the history when trying to fix the `indeterminate` issue, I'll note that @drewnoakes captured the timing issue working correctly once in this .gif - when unchecking Executable near the beginning of the animation, All shows the indeterminate state. But all other times unchecking things it does not.
